File: ColorCorrector/ColorCorrector/Plugin/HueSaturation.m
//
//  HueSaturation.m
//  ColorCorrector
//
//  Created by Joseph Slinker on 6/21/22.
//

#import "HueSaturation.h"

static NSString*    kKey_Hue        = @"Hue";
static NSString*    kKey_Saturation = @"Saturation";

@implementation HueSaturation

- (instancetype)initWithHue:(double)hue
                 saturation:(double)sat
{
    self = [super init];
    
    if (self != nil)
    {
        self.hue = hue;
        self.saturation = sat;
    }
    
    return self;
}

+ (BOOL)supportsSecureCoding
{
    return YES;
}

+ (instancetype)fromAPI: (id<PROAPIAccessing>)apiManager forParameter: (UInt32)parameter atTime: (CMTime)currentTime {
    // Retrieve the hue/saturation parameter
    id<FxParameterRetrievalAPI_v6> paramAPI = [apiManager apiForProtocol:@protocol(FxParameterRetrievalAPI_v6)];

    HueSaturation* hueSat  = nil;
    [paramAPI getCustomParameterValue:&hueSat fromParameter:parameter atTime:currentTime];
    return hueSat;
}

+ (void)setFromAPI: (id<PROAPIAccessing>)apiManager withID:(UInt32)parameterID hueRadians: (double)hueRadians saturation: (double)saturation atTime: (CMTime) currentTime {
    id<FxParameterSettingAPI_v5> paramAPI = [apiManager apiForProtocol:@protocol(FxParameterSettingAPI_v5)];
    HueSaturation* colorData = [[[HueSaturation alloc] initWithHue:hueRadians saturation:saturation]
                                       autorelease];
    [paramAPI setCustomParameterValue:colorData toParameter:parameterID atTime:currentTime];
}

+ (instancetype)fromPluginState: (NSData*)data {
    HueSaturation* hsv;
    [data getBytes:&hsv length:sizeof(hsv)];
    
    double hue = 0.0;
    double sat = 0.0;
    [data getBytes:&hue length:sizeof(hue)];
    
    NSRange satRange = NSMakeRange(sizeof(hue), sizeof(sat));
    [data getBytes:&sat range:satRange];
    
//    NSRange valueRange  = NSMakeRange(sizeof(hue) + sizeof(sat), sizeof(val));
//    [pluginState getBytes:&val
//                    range:valueRange];
    
    return [[HueSaturation alloc] initWithHue:hue saturation:sat];
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super init];
    
    if (self != nil)
    {
        _hue = [aDecoder decodeDoubleForKey:kKey_Hue];
        _saturation = [aDecoder decodeDoubleForKey:kKey_Saturation];
    }
    
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodeDouble:_hue
                  forKey:kKey_Hue];
    [aCoder encodeDouble:_saturation
                  forKey:kKey_Saturation];
}

- (instancetype)copyWithZone:(NSZone *)zone
{
    HueSaturation* newInstance = [[HueSaturation alloc] initWithHue:self.hue saturation:self.saturation];
    return newInstance;
}

- (NSObject<NSSecureCoding, NSCopying>*)interpolateBetween:(NSObject<NSSecureCoding, NSCopying>*)rightValue
                                                withWeight:(float)weight
{
    HueSaturation*    rhs = (HueSaturation*)rightValue;
    double            newHue  = fmod(self.hue + (rhs.hue - self.hue) * weight, M_PI * 2.0);
    double            newSat  = self.saturation + (rhs.saturation - self.saturation) * weight;
    HueSaturation*    result  = [[HueSaturation alloc] initWithHue:newHue
                                                            saturation:newSat];
    return [result autorelease];
}

- (BOOL)isEqual:(NSObject<NSSecureCoding, NSCopying>*)object
{
    HueSaturation*    rhs = (HueSaturation*)object;
    if ((self.hue == rhs.hue) && (self.saturation == rhs.saturation))
    {
        return YES;
    }
    
    return NO;
}

@end

File: ColorCorrector/ColorCorrector/Plugin/HueSaturation.swift
////
////  HueSaturation.swift
////  ColorCorrector
////
////  Created by Joseph Slinker on 4/27/22.
////
//
//import Foundation
//
//private enum CodingKey: String {
//    case Hue = "CodingKeyHue"
//    case Saturation = "CodingKeySaturation"
//}
//
//class HueSaturation: NSObject, NSSecureCoding, NSCopying, FxCustomParameterInterpolation_v2 {
//    
//    let hue: Double
//    let saturation: Double
//    
//    init(hue: Double, saturation: Double) {
//        self.hue = hue
//        self.saturation = saturation
//    }
//    
//    // MARK: - NSSecureCoding
//    
//    static var supportsSecureCoding: Bool = true
//    
//    // MARK: - NSCopying
//    
//    func copy(with zone: NSZone? = nil) -> Any {
//        return HueSaturation(hue: self.hue, saturation: self.saturation)
//    }
//    
//    // MARK: - NSCoding
//    
//    func encode(with coder: NSCoder) {
//        coder.encode(self.hue, forKey: CodingKey.Hue.rawValue)
//        coder.encode(self.saturation, forKey: CodingKey.Saturation.rawValue)
//    }
//    
//    required init?(coder: NSCoder) {
//        super.init()
//        self.hue = coder.decodeDouble(forKey: CodingKey.Hue.rawValue)
//        self.saturation = coder.decodeDouble(forKey: CodingKey.Saturation.rawValue)
//    }
//    
//    // MARK: - FxCustomParameterInterpolation_v2
//    
//    func interpolateBetween(_ rightValue: NSCopying & NSSecureCoding & NSObjectProtocol, withWeight weight: Float) -> NSCopying & NSSecureCoding & NSObjectProtocol {
//        let rhs = rightValue as! HueSaturation
//        let weight = Double(weight)
//        let newHue = fmod(self.hue + (rhs.hue - self.hue) * weight, Double.pi * 2.0)
//        let newSaturation = self.saturation + (rhs.saturation - self.saturation) * weight
//        return HueSaturation(hue: newHue, saturation: newSaturation)
//    }
//    
//    @objc func isEqual(_ rightValue: NSCopying & NSSecureCoding & NSObjectProtocol) -> Bool {
//        let rhs = rightValue as! HueSaturation
//        return self.hue == rhs.hue && self.saturation == rhs.saturation
//    }
//    
////    override func isEqual(_ object: Any?) -> Bool {
////        guard let rhs = object as? HueSaturation else { return false }
////        return self.hue == rhs.hue && self.saturation == rhs.saturation
////    }
//    
//}
